src/backend/optimizer/plan/createplan.c

```diff
@@ -6726,7 +6876,18 @@ adjust_modifytable_flow(PlannerInfo *root, ModifyTable *node, List *is_split_upd
 					node->action_col_idxes = lappend_int(node->action_col_idxes, -1);
 					node->ctid_col_idxes = lappend_int(node->ctid_col_idxes, -1);
 					node->oid_col_idxes = lappend_int(node->oid_col_idxes, 0);
-					request_explicit_motion(subplan, rti, root->glob->finalrtable);
+
+					/*
+					 * If an Explicit Motion has no Motions underneath it,
+					 * then the row to update must originate from the same
+					 * segment, and no Motion is needed.
+					 *
+					 * We elide the motion even if there are Motions, as long
+					 * as they are not between the scan on the target table
+					 * and the ModifyTable.
+					 */
+					if (!can_elide_explicit_motion(subplan, root->parse->rtable, rte->relid))
+						request_explicit_motion(subplan, rti, root->glob->finalrtable);
```

Comment:
Previously, we request explicit motions unconditionally here. So motion detection is implemented somewhere else. Why it was duplicated and left unchanged?

Reply:
FYI: The solution is based on [the commit from `7.x`](https://github.com/greenplum-db/gpdb/commit/a46400ef997e793e47ae7b8f1672e6ac29190925) and [a comment](https://github.com/greenplum-db/gpdb/blame/946eb7793cb6b35f395046f84af26021a815f165/src/backend/optimizer/plan/createplan.c#L3074-L3078). I couldn't backport the solution from `7.x` that uses paths and locuses (in [`create_motion_plan`](https://github.com/arenadata/gpdb/blob/ee5cbe6d1c276c77342ff749e42f7d3c573d8598/src/backend/optimizer/plan/createplan.c#L1337)) without breaking other motions, since there is no way to check if this is going to be specifically an `Explicit Redistribute Motion` for `UPDATE`/`DELETE` which can be safely elided.

`request_explicit_motion` does not add a motion by itself. [There are condition](https://github.com/arenadata/gpdb/blob/ee5cbe6d1c276c77342ff749e42f7d3c573d8598/src/backend/cdb/cdbmutate.c#L939-L942) that adds the motions only if there is any motions underneath. [The check itself](https://github.com/arenadata/gpdb/blob/ee5cbe6d1c276c77342ff749e42f7d3c573d8598/src/backend/cdb/cdbmutate.c#L1011-L1020) is a flag that is set to true if any of the visited nodes was a motion, while recursing the plan tree in similar way as in this patch.

Reply:
> The solution is based on [the commit from 7.x](https://github.com/greenplum-db/gpdb/commit/a46400ef997e793e47ae7b8f1672e6ac29190925) and [a comment](https://github.com/greenplum-db/gpdb/blame/946eb7793cb6b35f395046f84af26021a815f165/src/backend/optimizer/plan/createplan.c#L3074-L3078).

It's a good point to be remarked in PR description. But at the first glance, these patches have little in common because there was heavy refactoring in the gpdb 7.

> [The check itself](https://github.com/arenadata/gpdb/blob/ee5cbe6d1c276c77342ff749e42f7d3c573d8598/src/backend/cdb/cdbmutate.c#L1011-L1020) is a flag that is set to true if any of the visited nodes was a motion, while recursing the plan tree in a similar way as in this patch.

So we have a special flag invented for applying explicit redistribute motion and plan walking logic. Why we should reinvent it one more time in another place?